lib/inventory/controller.go

```diff
@@ -0,0 +1,302 @@
+	sshKeepAliveOk  testEvent = "ssh-keep-alive-ok"
+	sshKeepAliveErr testEvent = "ssh-keep-alive-err"
+
+	sshUpsertOk  testEvent = "ssh-upsert-ok"
+	sshUpsertErr testEvent = "ssh-upsert-err"
+
+	sshUpsertRetryOk  testEvent = "ssh-upsert-retry-ok"
+	sshUpsertRetryErr testEvent = "ssh-upsert-retry-err"
+
+	handlerStart = "handler-start"
+	handlerClose = "handler-close"
+)
+
+type controllerOpts struct {
+	serverKeepAlive time.Duration
+	testEvents      chan testEvent
+}
+
+type controllerOption func(c *controllerOpts)
+
+func withServerKeepAlive(d time.Duration) controllerOption {
+	return func(opts *controllerOpts) {
+		opts.serverKeepAlive = d
+	}
+}
+
+func withTestEventsChannel(ch chan testEvent) controllerOption {
+	return func(opts *controllerOpts) {
+		opts.testEvents = ch
+	}
+}
+
+// Controller manages the inventory control streams registered with a given auth instance. Incoming
+// messages are processed by invoking the appropriate methods on the Auth interface.
+type Controller struct {
+	store           *Store
+	auth            Auth
+	serverKeepAlive time.Duration
+	serverTTL       time.Duration
+	testEvents      chan testEvent
+	closeContext    context.Context
+	cancel          context.CancelFunc
+}
+
+func NewController(auth Auth, opts ...controllerOption) *Controller {
+	var options controllerOpts
+	for _, opt := range opts {
+		opt(&options)
+	}
+
+	if options.serverKeepAlive == 0 {
+		baseKeepAlive := apidefaults.ServerKeepAliveTTL()
+		// use 1.5x standard server keep alive since we use a jitter that
+		// shortens the actual average interval.
+		options.serverKeepAlive = baseKeepAlive + (baseKeepAlive / 2)
+	}
+
+	ctx, cancel := context.WithCancel(context.Background())
+	return &Controller{
+		store:           NewStore(),
+		serverKeepAlive: options.serverKeepAlive,
+		serverTTL:       apidefaults.ServerAnnounceTTL,
+		auth:            auth,
+		testEvents:      options.testEvents,
+		closeContext:    ctx,
+		cancel:          cancel,
+	}
+}
+
+// RegisterControlStream registers a new control stream with the controller.
+func (c *Controller) RegisterControlStream(stream client.UpstreamInventoryControlStream, hello proto.UpstreamInventoryHello) {
+	handle := newUpstreamHandle(stream, hello)
+	c.store.Insert(handle)
+	go c.handleControlStream(handle)
+}
+
+// GetControlStream gets a control stream for the given server ID if one exists (if multiple control streams
+// exist one is selected pseudorandomly).
+func (c *Controller) GetControlStream(serverID string) (handle UpstreamHandle, ok bool) {
+	handle, ok = c.store.Get(serverID)
+	return
+}
+
+// Iter iterates across all handles registered with this controller.
+// note: if multiple handles are registered for a given server, only
+// one handle is selected pseudorandomly to be observed.
+func (c *Controller) Iter(fn func(UpstreamHandle)) {
+	c.store.Iter(fn)
+}
+
+func (c *Controller) testEvent(event testEvent) {
+	if c.testEvents == nil {
+		return
+	}
+	c.testEvents <- event
```

Comment:
Since this is only used in tests, I prefer the blocking.  Definitely a personal taste thing, but if the events aren't being consumed for some reason, I'd rather see the test hit an explicit timeout rather than lose events.  Feels like that might result in unpredictable behavior that could be more work to debug.

I could add some mechanism for propagating overflow errors to the test explicitly, but I'm not sure if its worth the added complexity.